src/puppetlabs/jdbc_util/pglogical.clj

```diff
@@ -44,4 +45,96 @@
+  that database are up and running, :disabled if any subscription has been
+  turned off, :down if any connection has been severed (overrides :disabled),
+  :none if pglogical replication is not configured, or :unknown. Note that
+  :down may not be returned for some time or at all, depending on your
+  postgresql TCP settings."
+  [statuses]
+  (let [all-good? #(every? (partial = "replicating") %)
+        disabled? #(some (partial = "disabled") %)
+        down? #(some (partial = "down") %)]
+    (cond
+      (empty? statuses) :none
+      (all-good? statuses) :running
+      (down? statuses) :down
+      (disabled? statuses) :disabled
+      :else :unknown)))
+
+(defn replica-replication-status
+  "Given a DB connection for a pglogical replica node, returns
+  the status of data replication. Returns :running if all subscriptions for
+  that database are up and running, :disabled if any subscription has been
+  disabled, :down if any connection has been severed (overrides :disabled),
+  :none if pglogical replication is not configured, else :unknown. Note that
+  :down may not be returned for some time or at all, depending on your
+  postgresql TCP settings."
+  [db]
+  (if (has-pglogical-extension? db)
+    (-> (jdbc/query db ["SELECT status from show_subscription_status()"] {:row-fn :status})
+        (consolidate-replication-status))
+    :none))
+
+(defn consolidate-provider-status
+  [statuses]
+  (let [all-active? #(every? true? %)]
+    (case
+      (empty? statuses) :none
+      (all-active? statuses) :active
+      :else :inactive)))
+
+(defn provider-replication-status
+  "Given a DB connection for a pglogical provider node, query whether the
+  node's subscription is currently active or not. Returns :active or
+  :inactive, or :none if pglogical is not installed."
+  [db]
+  (if (has-pglogical-extension? db)
+    (jdbc/query db ["SELECT active FROM pg_replication_slots WHERE database = ?" (:db db)])
+    :none))
+
+(defn replication-status
+  "Given a DB connection for a pglogical node and the role of that node,
+  :source or :replica, query replication status. Returns a map reporting the
+  role and its status. Valid status values vary depending on the role; see
+  other replication-status functions for more information."
+  [db role]
+  (let [status (if (= :source role)
+                 (provider-replication status db)
+                 (replica-replication-status db))]
+    {:role role
+     :status status}))
+
+(defn replication-alert
+  "Produces an alert that can be used if replication is down. This is only a
+  function for localization reasons.
+
+  This function takes a service name and a replication state, which can be one
+  of: :disabled, :down, or :unknown. It then returns a human-readable message."
+  [service-name replication-state]
+  (let [construct-error (fn [message] {:severity "error"
+                                       :message message})]
+
+  (case replication-state
```

Comment:
Can put the map on the outside and the case on the inside to avoid these function calls